src/app/icd/client/ICDClientStorage.h

```diff
@@ -0,0 +1,95 @@
+/*
+ *
+ *    Copyright (c) 2023 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include "ICDClientInfo.h"
+#include <crypto/CHIPCryptoPAL.h>
+#include <lib/core/CHIPConfig.h>
+#include <lib/core/CHIPPersistentStorageDelegate.h>
+#include <lib/core/DataModelTypes.h>
+#include <lib/core/ScopedNodeId.h>
+#include <lib/support/CodeUtils.h>
+#include <lib/support/CommonIterator.h>
+#include <stddef.h>
+
+namespace chip {
+namespace app {
+
+/**
+ * The ICDClientStorage class is an abstract interface that defines the operations
+ * for storing, retrieving and deleting ICD client information in persistent storage.
+ */
+class ICDClientStorage
+{
+public:
+    using ICDClientInfoIterator = CommonIterator<ICDClientInfo>;
+
+    virtual ~ICDClientStorage() = default;
+
+    /**
+     * Iterate through persisted ICD Client Info
+     *
+     * @return A valid iterator on success. Use CommonIterator accessor to retrieve ICDClientInfo
+     */
+    virtual ICDClientInfoIterator * IterateICDClientInfo() = 0;
```

Comment:
What's the use case for this, given the existence of `ProcessCheckInPayload` as a toplevel API?

Probably OK for now, but if, as I suspect, we end up not using it, I suggest removing it from the API surface.

Reply:
sure. would decide whether to keep it after ICD integration is completed. thanks

Reply:
https://github.com/project-chip/connectedhomeip/issues/30720